opm/simulators/linalg/ISTLSolverEbos.hpp

```diff
@@ -364,35 +367,47 @@ namespace Opm
 #if HAVE_MPI
                     if (useWellConn_) {
                         using ParOperatorType = Dune::OverlappingSchwarzOperator<Matrix, Vector, Vector, Comm>;
-                        linearOperatorForFlexibleSolver_ = std::make_unique<ParOperatorType>(getMatrix(), *comm_);
-                        flexibleSolver_ = std::make_unique<FlexibleSolverType>(*linearOperatorForFlexibleSolver_, *comm_, prm_, weightsCalculator,
-                                                                               pressureIndex);
+                        auto op = std::make_unique<ParOperatorType>(getMatrix(), *comm_);
+                        using FlexibleSolverType = Dune::FlexibleSolver<ParOperatorType>;
+                        auto sol = std::make_unique<FlexibleSolverType>(*op, *comm_, prm_, weightsCalculator, pressureIndex);
+                        preconditionerForFlexibleSolver_ = &(sol->preconditioner());
+                        linearOperatorForFlexibleSolver_ = std::move(op);
```

Comment:
The reason is that the types are different. The flexibleSolver_ for example is a unique pointer to an AbstractSolverType, but the type of `sol` is something concrete, that includes now the type of the operator that is passed in. This is a consequence of having to use the operator rather than just matrix and vector types as template arguments for FlexibleSolver.

However, maybe this is a bit defensive, would simply:
```
flexibleSolver_ = std::make_unique<FlexibleSolverType>(...);
```
work fine? I admit I have not tried. If so, then this could be simplified to get rid of the moves.

Reply:
Looking more closely at this, I have concluded that the code is more or less as it has to be. While it is possible to store the objects created (`op` and `sol`) directly in their destination objects as sketched above, we are then not able to use their concrete types, only the more abstract ones. The creation of `sol` requires the concrete type of `op`, and extracting the `preconditionerForFlexibleSolver_` requires the concrete type of `sol`, so both must be dealt with in the current manner using std::move(). If it was not for the `preconditionerForFlexibleSolver_` we could have gotten away with directly assigning to `flexibleSolver_`, but we need it since calling `flexibleSolver_->preconditioner()` is no longer possible with `flexibleSolver_` being a more abstract InverseOperator.

Reply:
> we need [the dynamic type of `flexibleSolver_`] since calling `flexibleSolver_->preconditioner()` is no longer possible with `flexibleSolver_` being a more abstract `InverseOperator`.

Well, you **could** down-cast to the appropriate (and known) type
```C++
dynamic_cast<FlexibleSolverType*>(this->flexibleSolver_.get())->preconditioner();
```
but, well, ..., ick.